### DISCUSSION TITLE
[cd-on-quit] How about skip redundant `cd` commands?

### DISCUSSION THREAD

**Discussion by rennsax** [jarun/nnn#1833] [Ideas]
When `NNN_TMPFILE` is set, `nnn` will write the `cd` command into the tmp file:

https://github.com/jarun/nnn/blob/f71b1309a9f304759b2452186538dc142da9b1a2/src/nnn.c?plain=1#L2843-L2846

But even if the current directory doesn't change, the `cd` command will still be executed, which make _duplicated entries in the directory stack_ of some shells (like zsh).

How about skip this kind of `cd` by conditionally writing the command?

I realized it in [my personal fork](https://github.com/rennsax/nnn/blob/af3cfb5b27b8c99f613ba382fa1e8996c86d50d5/src/nnn.c?plain=1#L2843-L2846):

```c
dprintf(fd, "[ \"$PWD\" = %s ] || cd %s", g_buf, g_buf);
```

But I know it's not portable among prevalent shells like fish.

**Comment by N-R-K** [comment]
You should do this in your quit-cd script instead of in `nnn.c`.

In fact, `nnn` shouldn't even write the cd command to begin with and it would've been ideal if we simply wrote the path as is into the file and nothing else. But the cd command was kept to keep existing scripts working.

**Reply by N-R-K** [reply]
> You should do this in your quit-cd script instead of in `nnn.c`.

To elaborate, I have the following in my own quit-cd script:

```bash
# vim: filetype=bash
n3() {
	[ "${NNNLVL:-0}" -ge 1 ] && exit # block nesting

	LESS="${LESS/F/}" nnn "$@"; local ec=$?

	if [ "$ec" -eq 0 ] && [ -f "$NNN_TMPFILE" ]; then
		local to=$(sed "s|^cd '||;s|'$||;s|'\\\''|'|g;" "$NNN_TMPFILE")
		[ "$to" != "$PWD" ] && cd "$to"
	fi
	rm -f "$NNN_TMPFILE" >/dev/null
}
```

This avoids `cd`-ing when directory hasn't changed. Additionally, the `$ec` check makes it avoid changing dir when I quit with `Q` (non-zero exit code).

I use zsh but I think this should work on bash too.
